### PR TITLE
Path iterator

### PR DESCRIPTION
**Pull request by dnfield**
This patch exposes most of `SkPath::Iter` in the engine.  This allows more inspection of `Path` objects - e.g. if you get a path, you could get all the necessary information to serialize it.  

Compared to `PathMetrics` (aka PathMeasure), it gives the actual points/verbs/weights used by Skia for the segments of a path (instead of their lengths, subpaths, and tangents).

The basic use case is "I want to be able to serialize a Path" - which could then be persisted and deserialized, or converted to some other format (e.g. a potential Flutter vector format, or SVG Path data).

Although there is no code currently in the engine that uses `SkPath::Iter` directly, it is used indirectly by other calls already.  At the bottom of it all, this code is really just exposing the verb and point arrays, and the weights for conic segments.

### CODE REVIEW

**Comment by Hixie** [comment]
I'm concerned that exposing this limits us dramatically in what we can do with Path in the future. For example, it requires that we not optimise the creation of Paths as they are constructed (e.g. if someone did two lineTo calls that could be turned into one, we couldn't do it anymore, because that would be detectable in this API and would thus mean that such changes were potentially breaking changes).

Given that any Path in the engine had to come from someone creating the Path in the first place, couldn't whoever created the Path take note of how they created it?

**Comment by dnfield** [comment]
As far as I understand, Skia can already do something like what you're suggesting.  This API doesn't guarantee that it will give you exactly what was put in, just that it will give you something that recreates what would be drawn.  Very small segments, for example, would be dropped.

**Comment by dnfield** [comment]
And I suppose it would be possible (perhaps even by a proxy path object), but it adds a layer of complication and may miss out on optimizations that Skia can already do or will do.

**Comment by Hixie** [comment]
Whether we say that the API will do such optimisations or not, people will depend on it returning the same thing every time (unless we randomize it somehow).

**Comment by Hixie** [comment]
What I learnt from the Web is that regardless of how much you document something, people will rely on exactly what you do anyway, even if they don't mean to. People rely on all kinds of crazy things, e.g. they'll write code with race conditions that rely on one operation taking longer than another such that when you optimize the first, their code breaks. The user equivalent of this is https://xkcd.com/1172/.

With APIs like these, you'll end up with people writing vector image editing tools that try to round-trip files through Path objects, and some optimization in Skia next year that's dependent on the precise device pixel ratio will cause certain subpaths to be skipped which will result in the tool losing data, except nobody will notice until 6 months after the app has shipped with the new engine, and by then lots of people will have damaged their images. (That's just a crazy example I just came up with, who knows what the real problem will be.)

What's the use case you're trying to actually address?

**Comment by Hixie** [comment]
So the use case is creating a format that is faster than SVG?

I don't see why we would need this for that purpose. The part that is needed to do that is a deserializer; this is a serializer.

**Comment by dnfield** [comment]
The idea is it could all be done in flutter, and be useful for saving path input taken in from other sources (e.g. from gestures for painting or signing).  But I suppose this could still be done with a wrapper around path or a proxy path object too.

**Comment by Hixie** [comment]
Yeah if you just want to record a signature then doing it in a custom object that can then output a Path seems like a much more robust solution.

**Comment by Hixie** [comment]
(in particular, that would let you do things like hold timing and pressure information, which for a signature might be important)

**Comment by dnfield** [comment]
Alright.  I think I could use some of this as an add on to the path drawing module I have and handle it on the dart end. My drive on this was that it would be nice to have components that don't need to wrap paths, but I see your point here - keeping the core object opaque allows for more flexibility going forward.

**Comment by DeRealMorgan** [comment]
If people dont read the docs or ignore how to use a piece of code then it is not the frameworks fault. 
Programmers should be educated enough to use those features.

An idea I am having:
The main concern apparently is that there could be hidden optimizations in the background that will change how paths are created while the developer isnt aware of that.

Why dont do both: Leave the path as is but add a boolean in the constructor "useRawData" which leaves the paths as it is at the moment for the normal user.

Users with advanced usecases can set this boolean to true - leading to no (or unsignifact) optimizations but also exposing the "rawData" to the developer.

I can only talk for myself but I am handling the path optimization myself (as I want low level control)

File: testing/dart/pubspec.yaml [hunk 1]
```diff
@@ -1,4 +1,7 @@
 name: engine_tests
 dependencies:
-  test: 0.12.40
-  path: 1.6.0
+  test: 0.12.41
+  path: 1.6.1
+dev_dependencies:
+  flutter_test:
+    sdk: flutter
```
Comment (dnfield) [inline testing/dart/pubspec.yaml:7]: Note to self: this was just wrong, don't do this.

**Comment by skquo** [comment]
Finally sounds good, @dnfield, when will iterator be integrated to the engine?

**Comment by dnfield** [comment]
This has been in for over two years now...

**Comment by skquo** [comment]
Sorry, I don't get it? When are we able to use PathIterator from Flutter ui library?

**Comment by dnfield** [comment]
Oh! Sorry, I got confused about which patch this was and thought you were talking about PathMetrics. This patch never landed. There's some more discussion about it at https://github.com/flutter/flutter/issues/29125. Part of the challenge is coming up with a really compelling use case for adding it.

**Pull request closed without merging**